[Security Solution] Add alert and cloud insights to document flyout

christineweng opened the pull request:
## Summary

This PR adds alert count, misconfiguration and vulnerabilities insights to alert/event flyout. If data is not available, the insights are hidden.

[Mocks](https://www.figma.com/design/ubvhBGHee58diJNvSiy0GZ/%5B8.%2B%5D-%5BAlerts%5D-Expandable-Event-Flyout?node-id=8017-179782&node-type=canvas&t=0YjHfPi9zOUFUScc-0)

![image](https://github.com/user-attachments/assets/ba706ab8-448a-4286-8229-c4c398136638)

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

<!--ONMERGE {"backportTargets":["8.15","8.x"]} ONMERGE-->

<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

christineweng committed:
add misconfigurations and vulnerabilities

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

elasticmachine commented:
Pinging @elastic/security-threat-hunting-investigations (Team:Threat Hunting:Investigations)

christineweng committed:
add alert insights

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/240489)
* Commit: 2ce61babc860a5a3018acab944f5f9cba3bf36cc

### Failed CI Steps
* [Defend Workflows Cypress Tests on Serverless #8](https://buildkite.com/elastic/kibana-pull-request/builds/240489#01926e7a-e190-4a98-b0a8-9e4e746887a0)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/240489#01926e7a-e190-4a98-b0a8-9e4e746887a0) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/240489/jobs/01926e7a-e190-4a98-b0a8-9e4e746887a0/artifacts/01926ea7-bb07-49de-9af3-808ca4f89bfb) Defend Workflows Cypress Tests on Serverless #<span></span>8 / User Roles for Security Complete PLI with Endpoint Complete addon for role: t1_analyst "before all" hook for "should have READ access to Endpoint list page" "before all" hook for "should have READ access to Endpoint list page"

cc @christineweng

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240489","number":240489,"commit":"2ce61babc860a5a3018acab944f5f9cba3bf36cc"}],"number":240489}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/240873)
* Commit: 247008126f9b1c6b7cf3372f42b0c8044fa17e21

### Failed CI Steps
* [Build Storybooks](https://buildkite.com/elastic/kibana-pull-request/builds/240873#019272a8-5d20-4055-814c-17f4256d2522)

### History
* :broken_heart: [Build #240489](https://buildkite.com/elastic/kibana-pull-request/builds/240489) failed 2ce61babc860a5a3018acab944f5f9cba3bf36cc

cc @christineweng

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240873","number":240873,"commit":"247008126f9b1c6b7cf3372f42b0c8044fa17e21"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240489","number":240489,"commit":"2ce61babc860a5a3018acab944f5f9cba3bf36cc"}],"number":240873}
buildkite-pr-comment-->

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/document_details/shared/components/alert_count_insight.tsx:
```diff
@@ -0,0 +1,85 @@
+  fieldName,
+  direction,
+  'data-test-subj': dataTestSubj,
+}: {
+  name: string;
+  fieldName: 'host.name' | 'user.name';
+  direction?: EuiFlexGroupProps['direction'];
+  ['data-test-subj']?: string;
+}) => {
```
would you be ok extracting these into a `AlertCountInsightsProps` interface and have documentation for each props?

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/document_details/shared/components/alert_count_insight.tsx:
```diff
@@ -0,0 +1,85 @@
+  const uniqueQueryId = useMemo(() => `${ENTITY_ALERT_COUNT_ID}-${uuid()}`, []);
+  const entityFilter = useMemo(() => ({ field: fieldName, value: name }), [fieldName, name]);
+
+  const { items, isLoading } = useSummaryChartData({
```
awesome job remembering to reuse this hook!

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/document_details/shared/components/insight_distribution_bar.tsx:
```diff
@@ -0,0 +1,91 @@
+        </EuiText>
+      </EuiFlexItem>
+      <EuiFlexItem>
+        <EuiFlexGroup gutterSize="xs">
```
let add `responsive={false}` here so we guarantee this look
![Screenshot 2024-10-09 at 4 26 05 PM](https://github.com/user-attachments/assets/3dc213a8-28d2-4827-bbb2-bbcf13f7eaf3)
instead of this one
![Screenshot 2024-10-09 at 4 26 28 PM](https://github.com/user-attachments/assets/b1ce2bbc-f2e8-4154-9c66-3c6cc153ff07)

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/document_details/shared/components/alert_count_insight.tsx:
```diff
@@ -0,0 +1,85 @@
+
+const ENTITY_ALERT_COUNT_ID = 'entity-alert-count';
+
+/*
+ * Alert count insight in the alert/event flyout.
+ */
```
Would you be ok adding a bit more information here? I'm thinking it'd be nice to specify that the count will always reflect the `ciritcal` alerts only. I know it's in the code but having it here would make a bit more obvious?

PhilippeOberti commented on x-pack/plugins/security_solution/public/cloud_security_posture/components/vulnerabilities/vulnerabilities_insight.tsx:
```diff
@@ -0,0 +1,80 @@
+  hostName,
+  direction,
+  'data-test-subj': dataTestSubj,
+}: {
+  hostName: string;
+  direction?: EuiFlexGroupProps['direction'];
+  ['data-test-subj']?: string;
```
extract these into a `VulnerabilitiesInsightProps` interface and docs each prop?

PhilippeOberti commented on x-pack/plugins/security_solution/public/cloud_security_posture/components/vulnerabilities/vulnerabilities_insight.tsx:
```diff
@@ -0,0 +1,80 @@
+import { getVulnerabilityStats, hasVulnerabilitiesData } from '@kbn/cloud-security-posture';
+import { InsightDistributionBar } from '../../../flyout/document_details/shared/components/insight_distribution_bar';
+
+/*
+ * Vulnerabilities insight in the alert/event flyout.
+ */
```
Similar here, would you be ok enhancing this docs by saying that we display the critical vulnerabilities in the count next to the chart?

PhilippeOberti commented on x-pack/plugins/security_solution/public/cloud_security_posture/components/vulnerabilities/vulnerabilities_insight.tsx:
I know we discussed about this, but I would consider moving this to the flyout folder, along side `alert_count_insight`. I feel like it makes sense especially because this is using the `InsightDistrubutionBar` component, which is there and we own.
I feel like we can add enough unit tests and make sure that everything is typed correctly to protect us from changes that would break the UI in the flyout.

I'll think about it a bit more...

PhilippeOberti commented on x-pack/plugins/security_solution/public/cloud_security_posture/components/misconfiguration/misconfiguration_insight.tsx:
```diff
@@ -0,0 +1,66 @@
+import { InsightDistributionBar } from '../../../flyout/document_details/shared/components/insight_distribution_bar';
+import { getFindingsStats } from './misconfiguration_preview';
+
+/*
+ * Misconfigurations insight in the alert/event flyout.
+ */
```
same here, would you be ok adding a bit to this documentation, saying that we show the count of failed misconfigurations in the label on the right of the distribution chart?

PhilippeOberti commented on x-pack/plugins/security_solution/public/cloud_security_posture/components/misconfiguration/misconfiguration_insight.tsx:
```diff
@@ -0,0 +1,66 @@
+  fieldName,
+  direction,
+  'data-test-subj': dataTestSubj,
+}: {
+  name: string;
+  fieldName: 'host.name' | 'user.name';
+  direction?: EuiFlexGroupProps['direction'];
+  ['data-test-subj']?: string;
```
extract to `MisconfigurationsInsightProps` and add docs?

PhilippeOberti commented on x-pack/plugins/security_solution/public/cloud_security_posture/components/misconfiguration/misconfiguration_insight.tsx:
same here, to me right now it makes more sense to have it part of the flyout folder but I'll think about it more and we can discuss again

PhilippeOberti reviewed (commented):
this looks and works amazing, awesome job! I left a few comments, let me know what you think!

christineweng committed:
PR Comments

christineweng commented:
@PhilippeOberti Thanks for all the feedback! I have updated the props and doc strings. Also moved the components into our flyout folder. I agree it is okay to stay in our folder, we have many components that we don't technically own the backend for (assistant, assignee etc.)

christineweng commented:
## Files by Code Owner

### elastic/kibana-cloud-security-posture

* x-pack/packages/security-solution/distribution_bar/src/distribution_bar.stories.tsx
* x-pack/packages/security-solution/distribution_bar/src/distribution_bar.test.tsx
* x-pack/packages/security-solution/distribution_bar/src/distribution_bar.tsx
* x-pack/plugins/security_solution/public/cloud_security_posture/components/misconfiguration/misconfiguration_preview.tsx

### elastic/security-threat-hunting-investigations & elastic/security-solution

* x-pack/plugins/security_solution/public/flyout/document_details/left/components/host_details.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/left/components/host_details.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/left/components/test_ids.ts
* x-pack/plugins/security_solution/public/flyout/document_details/left/components/user_details.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/left/components/user_details.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/host_entity_overview.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/host_entity_overview.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/test_ids.ts
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/user_entity_overview.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/user_entity_overview.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/alert_count_insight.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/alert_count_insight.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/insight_distribution_bar.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/insight_distribution_bar.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/misconfiguration_insight.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/misconfiguration_insight.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/test_ids.ts
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/vulnerabilities_insight.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/shared/components/vulnerabilities_insight.tsx

PhilippeOberti reviewed (approved):
thanks for making all the changes. Code looks great!

I left one last comment regarding the count shown next to the distribution chart. This is a non blocker (and could even be not desired)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/240937)
* Commit: 382b23ca5b4be890f9d7fab0a10150c93ff03971

### Failed CI Steps
* [Build Storybooks](https://buildkite.com/elastic/kibana-pull-request/builds/240937#0192737c-9b20-4202-b554-994a471150c8)

### History
* :broken_heart: [Build #240873](https://buildkite.com/elastic/kibana-pull-request/builds/240873) failed 247008126f9b1c6b7cf3372f42b0c8044fa17e21
* :broken_heart: [Build #240489](https://buildkite.com/elastic/kibana-pull-request/builds/240489) failed 2ce61babc860a5a3018acab944f5f9cba3bf36cc

cc @christineweng

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240937","number":240937,"commit":"382b23ca5b4be890f9d7fab0a10150c93ff03971"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240873","number":240873,"commit":"247008126f9b1c6b7cf3372f42b0c8044fa17e21"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240489","number":240489,"commit":"2ce61babc860a5a3018acab944f5f9cba3bf36cc"}],"number":240937}
buildkite-pr-comment-->

christineweng commented on x-pack/plugins/security_solution/public/flyout/document_details/shared/components/insight_distribution_bar.tsx:
```diff
@@ -0,0 +1,91 @@
+  const { euiTheme } = useEuiTheme();
+  const xsFontSize = useEuiFontSize('xs').fontSize;
+
+  const countStr = useMemo(() => {
+    return count >= COUNT_LIMIT ? `${COUNT_LIMIT}+` : `${count}`;
+  }, [count]);
```
Let me check with Fernanda, 999+ comes from the mock
![image](https://github.com/user-attachments/assets/7a2172ff-8608-4e32-a350-530378e2ff97)

christineweng committed:
update count

christineweng committed:
Merge branch 'main' into misconfig-in-alert-flyout

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/241293)
* Commit: 97c4183080c53b1678ad1f157fecfe99e7bde087

### Failed CI Steps
* [Investigations - Security Solution Cypress Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/241293#01927742-101e-4b73-9f7d-02a12f99fcda)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/241293#01927742-101e-4b73-9f7d-02a12f99fcda) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/241293/jobs/01927742-101e-4b73-9f7d-02a12f99fcda/artifacts/01927779-d7fb-4acc-8786-4ce1c5b678b0) Investigations - Security Solution Cypress Tests #<span></span>7 / Multiple indicators Indicator page search should handle all search actions should handle all search actions

### History
* :broken_heart: [Build #240937](https://buildkite.com/elastic/kibana-pull-request/builds/240937) failed 382b23ca5b4be890f9d7fab0a10150c93ff03971
* :broken_heart: [Build #240873](https://buildkite.com/elastic/kibana-pull-request/builds/240873) failed 247008126f9b1c6b7cf3372f42b0c8044fa17e21
* :broken_heart: [Build #240489](https://buildkite.com/elastic/kibana-pull-request/builds/240489) failed 2ce61babc860a5a3018acab944f5f9cba3bf36cc

cc @christineweng

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/241293","number":241293,"commit":"97c4183080c53b1678ad1f157fecfe99e7bde087"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240937","number":240937,"commit":"382b23ca5b4be890f9d7fab0a10150c93ff03971"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240873","number":240873,"commit":"247008126f9b1c6b7cf3372f42b0c8044fa17e21"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240489","number":240489,"commit":"2ce61babc860a5a3018acab944f5f9cba3bf36cc"}],"number":241293}
buildkite-pr-comment-->

christineweng committed:
Merge branch 'main' into misconfig-in-alert-flyout

kibanamachine commented:
Starting backport for target branches: 8.15, 8.x

https://github.com/elastic/kibana/actions/runs/11281745729

kibanamachine commented:
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.15|Backport failed because of merge conflicts|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/195825">](https://github.com/elastic/kibana/pull/195825)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 195509
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/11283760507

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.x|Cherrypick failed because the selected commit (cd217c07) is empty. It looks like the commit was already backported in https://github.com/elastic/kibana/pull/195825|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 195509
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
